analytics-integrations/kahuna/src/main/java/com/segment/analytics/internal/integrations/KahunaIntegration.java

```diff
@@ -57,25 +81,64 @@
   @Override public void identify(IdentifyPayload identify) {
     super.identify(identify);
 
-    String username = identify.traits().username();
-    KahunaAnalytics.setUsernameAndEmail(isNullOrEmpty(username) ? identify.userId() : username,
-        identify.traits().email());
-
-    KahunaAnalytics.setUserCredential(USERNAME_KEY, identify.traits().username());
-    KahunaAnalytics.setUserCredential(EMAIL_KEY, identify.traits().email());
-    KahunaAnalytics.setUserCredential(FACEBOOK_KEY, identify.traits().getString("facebook"));
-    KahunaAnalytics.setUserCredential(TWITTER_KEY, identify.traits().getString("twitter"));
-    KahunaAnalytics.setUserCredential(LINKEDIN_KEY, identify.traits().getString("linkedin"));
-
-    KahunaAnalytics.setUserAttributes(identify.traits().toStringMap());
+    KahunaAnalytics.setUserCredential(USER_ID_KEY, identify.userId());
+
+    Traits identityTraits = identify.traits();
+    Map<String, String> userAttributes = new HashMap<>();
+    for (String key : identityTraits.keySet()) {
+      if (identityTraits.get(key) == null) {
+        continue; // Skip null value objects.
+      } else if (SUPPORTED_KAHUNA_CREDENTIALS.contains(key)) {
+        KahunaAnalytics.setUserCredential(key, identityTraits.getString(key));
+      } else {
+        // We'll track unsupported Kahuna User Credentials as Kahuna User Attributes instead
+        // and make sure to format any Date objects appropriately.
+        Object value = identityTraits.get(key);
+        if (value instanceof Date) {
+          userAttributes.put(key, Utils.toISO8601Date((Date) value));
+        } else {
+          userAttributes.put(key, String.valueOf(value));
+        }
+      }
+    }
+    KahunaAnalytics.setUserAttributes(userAttributes);
```

Comment:
For the categories_viewed attribute, we add to the existing list that the SDK knows (a comma separated list). In here, we are tracking new keys (or overriding the values those key reference) and our Mobile SDK takes care of merging the attributes internally and letting the server know of any delta changes as needed. For these attributes we don't care about the previous values, only the new ones so we don't need to get the previous values.